Add JavaScript MIPs for scaling textures larger and smaller

ktbee opened the pull request:
### Resolves

https://github.com/LLK/scratch-render/issues/374, potentially also https://github.com/LLK/scratch-render/issues/334

### Proposed Changes

[As discussed on other issues](https://github.com/LLK/scratch-render/issues/334), adding mipmap functionality will be helpful for scaling images at a good resolution in Scratch. This PR:

- Adds a SVGMIP as a wrapper for WebGL textures to handle creating them at a certain scale and updating them if needed
- Stores JavaScript mips on SVGSkin in an array `_scaledMIPs` using an index offset to allow for mips that have a scale of less than one. We're using an array here for better performance instead of an object with more descriptive keys
- Extends while loop in `SVGSkin.getTexture` to get smaller scales in addition to larger ones

Here is an example repro project. After the green flag click, the original sprite is scaled up and down while the clone remains unscaled. Both have a better resolution because they no longer share a texture at all times and use the one at its requested scale. https://scratch.mit.edu/projects/295219677/editor/

The screenshots' differences are a little tough to see at a smaller size, but it's clear when you click them and see them in a new tab. Also these were taken on a lower resolution monitor. If your screen has a hidpi, these differences are harder to see. 

| Production     | PR Changes |
| ----------- | ----------- |
| ![Screen Shot 2019-04-01 at 2 28 15 PM](https://user-images.githubusercontent.com/7991694/55350589-719f5c00-548a-11e9-8046-99b480af4fde.png) |  ![Screen Shot 2019-04-01 at 2 25 26 PM](https://user-images.githubusercontent.com/7991694/55350453-238a5880-548a-11e9-84fb-7ceff3e45116.png) |

### Memory usage

[In our previous discussion of this topic,](https://github.com/LLK/scratch-render/issues/334#issuecomment-449143929) one point against scaling mips with JavaScript instead of automatically on the GPU is that this method would use more memory. From some profiling I did with [Chrome](https://www.lifewire.com/google-chrome-task-manager-4103619) and [Firefox](https://developer.mozilla.org/en-US/docs/Mozilla/Performance/about:memory), it seems like [this test project](https://scratch.mit.edu/projects/295219677/editor/) hovers around the same amount of memory on the `develop` branch and this feature branch. Here are the results I found:

| **Chrome**  | `develop`      | `javascript-scaled-textures` |
| ----------- | ----------- | ----------- |
| **Memory Footprint** | 192 MB |  190 MB  |
| **GPU Memory**  |  28.3 MB      |    28.5 MB |

| **Firefox**  | `develop`      | `javascript-scaled-textures` |
| ----------- | ----------- | ----------- |
|Trial 1  |  81.96 MB (09.74%)  | 82.49 MB (09.79%) |
|Trial 2  | 73.19 MB (08.73%)  | 104.62 MB (11.86%) |
|Trial 3  | 99.08 MB (11.43%) | 84.35 MB (09.85%) |


Theoretically there should be a memory usage increase because there are more textures. I added the surface areas of power of 2 sized images between 2x2 and 1024x1024 and found that they would use about 33% more memory than just a single 1024x1024 image, but perhaps in the way we're using them in this branch, the browser is able to release some of that memory. From what I can see using Chrome and Firefox's tools, multiple JavaScript scaled textures hasn't had much of an impact on a tab's memory usage. I'm curious to hear more thoughts on this though!

### What this hasn't solved yet

I've found this implementation of mips successfully scales sprites that have been changes by a "looks" block, but sprites that haven't been scaled by a block still don't scale well. For example, when you reduce the browser's width while using the editor, the stage scales down and the sprite's quality is reduced:

| Full size stage      | Smaller stage |
| -----------            | ----------- |
| ![Screen Shot 2019-04-01 at 2 17 11 PM](https://user-images.githubusercontent.com/7991694/55350065-318ba980-5489-11e9-8cf8-ce7567154d1c.png)     | ![Screen Shot 2019-04-01 at 2 17 23 PM](https://user-images.githubusercontent.com/7991694/55350080-3ea89880-5489-11e9-89a9-d4d04e2dc2a3.png) |

Fullscreen mode is another example of this:

![Screen Shot 2019-04-01 at 2 32 13 PM](https://user-images.githubusercontent.com/7991694/55350862-0609be80-548b-11e9-980f-92b8e58b1f31.png)

Thoughts on how to scale sprites in this scenario and also thoughts on this approach for mipmaps are much appreciated!

thisandagain commented:
/cc @kchadha @rschamp @paulkaplan

adroitwhiz commented:
> Did you consider using OpenGL's built-in mipmap support? There might be a reason not to choose this that I haven't thought of, but it seems like it would simplify the code.

IIRC, OpenGL mipmaps only work with textures whose dimensions are a power of 2.

ktbee commented:
We discussed this a bit offline, but just to follow up for anyone watching the issue: 

@adroitwhiz has a good point, for WebGL 1, automatic mipmaps only work for images with dimensions that are a power of two. [We'd need to use an OpenGL extension or upgrade our WebGL version to be able to automatically generate mipmaps for non-power of two images.](https://www.khronos.org/webgl/wiki/WebGL_and_OpenGL_Differences#Non-Power_of_Two_Texture_Support)

ktbee commented:
Also @cwillisf @kchadha, I turned off twgl's automatic mipmap creation option in this PR and it is all set to be reviewed!

ktbee commented:
Hooray for tests! They helped me catch a couple bugs, which I fixed with the following changes:

- Use `this._texture.getTexture()` instead of just `this._texture` in SVGSkin's dispose function.
- Updated the silhouette for sprites with a scale of 1 to make sure it's set initially even if the sprite isn't scaled. 

@cwillisf let me know what you think!

adroitwhiz commented:
Shouldn't `SVGSkin.dispose` dispose of *all* `SVGMIP` textures, instead of just the current one?

adroitwhiz commented on src/SVGSkin.js:
```diff
@@ -25,19 +26,22 @@ class SVGSkin extends Skin {
     dispose () {
         if (this._texture) {
-            this._renderer.gl.deleteTexture(this._texture);
+            this._renderer.gl.deleteTexture(this._texture.getTexture());
```
Shouldn't `SVGSkin.dispose` call `gl.deleteTexture` on *all* `SVGMIP` textures, instead of just the current one?

ktbee replied:
Oh good point! I've updated `dispose` to delete all of the textures and reset `_scaledMIPs` and `_largestTextureScale`. Let me know if I'm missing anything else.

adroitwhiz commented on src/SVGSkin.js:
```diff
@@ -1,9 +1,10 @@
+const SVGMIP = require('./SVGMIP');
 const SvgRenderer = require('scratch-svg-renderer').SVGRenderer;
 
 const MAX_TEXTURE_DIMENSION = 2048;
```
Could `gl.getParameter(gl.MAX_TEXTURE_SIZE)` potentially be used for this?

adroitwhiz commented on src/SVGSkin.js:
```diff
@@ -60,6 +70,34 @@ class SVGSkin extends Skin {
+                this._silhouette.update(textureData);
+            }
+
+            if (scale === 1) {
```
This seems a bit flaky-- it looks like it assumes `createMIP` will first be called with scale 1, which may not be the case. Could this be moved elsewhere, perhaps to `setSVG`?

ktbee replied:
Hmm, it's not meant to assume `createMIP` will be called with a scale 1. The intention instead is _if_ it gets called with a scale of 1, use the `textureData` to set the `_maxTextureScale` and update the silhouette. We need it to be in the callback in `createMIP` instead of `setSVG` because we need the `textureData` that is passed to the callback in `SVGMIP.draw`.

fsih replied:
I did this sequence of steps and this one weird trick happened
![line cat](https://user-images.githubusercontent.com/2855464/65554851-c29f8280-def8-11e9-8740-9e919865348e.gif)

fsih replied:
This thing I saw may have been https://github.com/LLK/scratch-gui/issues/5215

mzgoddard commented on src/SVGSkin.js:
```diff
@@ -1,9 +1,10 @@
 
 const MAX_TEXTURE_DIMENSION = 2048;
+const MIN_TEXTURE_SCALE = 1 / 256;
+const INDEX_OFFSET = 8;
```
How about:

```js
/**
 * All scaled renderings of the SVG are stored in an array. The 1.0 scale of
 * the SVG is stored at the 8th index. The smallest possible 1 / 256 scale
 * rendering is stored at the 0th index.
 * @const {number}
 */
```

mzgoddard commented on src/SVGSkin.js:
```diff
@@ -25,20 +26,29 @@ class SVGSkin extends Skin {
 
         /** @type {Number} */
-        this._maxTextureScale = 0;
+        this._largestTextureScale = 1;
```
Combined with another comment, we could get rid of _largestTextureScale. It is needed if we keep already created MIPs. If we dump all MIPs on setSVG calls, we don't need _largestTextureScale.

mzgoddard commented on src/SVGSkin.js:
```diff
@@ -100,43 +137,25 @@ class SVGSkin extends Skin {
-            this.setRotationCenter.apply(this, rotationCenter);
-            this.emit(Skin.Events.WasAltered);
-        });
+        if (typeof rotationCenter === 'undefined') rotationCenter = this.calculateRotationCenter();
```
In a different change we realized that setting the rotationCenter before the image has drawn will cause a "flash" of the old image at the wrong scale. We need to setRotationCenter after drawing.

One way we might do that is move the rotationCenter set logic in createMIP?

mzgoddard commented on src/SVGSkin.js:
```diff
@@ -100,43 +137,25 @@ class SVGSkin extends Skin {
-                this._texture = twgl.createTexture(gl, textureOptions);
-                this._silhouette.update(textureData);
-            }
+        this._svgRenderer.loadString(svgData);
```
We could:
1. Call dispose on each current MIPs.
2. Create a new array.
3. Set the 1.0 scale MIP at INDEX_OFFSET.

mzgoddard commented on src/SVGMIP.js:
```diff
@@ -0,0 +1,69 @@
+        this._callback = callback;
+        this.dirty = false;
+
+        this.draw();
```
It doesn't have to be for every MIP. But we should at least call draw in setSVG on the 1.0 scale MIP. This "primes" the SVGRenderer by loading an Image instance and drawing it to a canvas. The loading part can need a lot of time, so we should load the svg during the initial project load to avoid potentially large draw times when you switch to a SVG costume that hasn't been drawn before.

mzgoddard commented on src/SVGSkin.js:
```diff
@@ -60,11 +75,34 @@ class SVGSkin extends Skin {
+                this._silhouette.update(textureData);
+            }
+
+            if (resetCallback) resetCallback();
```
Run resetCallback before checking scaledMIPs.length. As is, this is a race condition. If you change the svgData, the second time, scaledMIPs may have an index higher than 8 (the 1.0 index) set. During that second change the silhouette will not update for the 1.0 scale MIP.

ktbee replied:
Makes sense! Thanks for catching that.

mzgoddard commented on src/SVGSkin.js:
```diff
@@ -60,11 +75,34 @@ class SVGSkin extends Skin {
+        const textureCallback = textureData => {
+            // Check if we have the largest MIP
+            // eslint-disable-next-line no-use-before-define
+            if (!this._scaledMIPs.length || this._scaledMIPs[this._scaledMIPs.length - 1]._scale <= scale) {
```
I think we can make this a little simpler.

One way would be if you pass another argument to createMIP. Pass the scaled index and we can `this._scaledMIPs.length - 1 === scaleIndex` to test for this.

mzgoddard commented on src/SVGSkin.js:
```diff
@@ -73,80 +111,70 @@ class SVGSkin extends Skin {
-                super.setEmptyImageData();
-                return;
-            }
+    resetMIPs (mip, rotationCenter) {
```
I think this method is pretty slick.

BryceLTaylor commented:
Test Plan:

Description: We will now take vector images and render them at several different sizes in memory to make scaling images look better.

To test:
* clone the sprite so you can see it and the original and try all of the below (make changes to both)
* Take a vector image and scale it to different sizes
* Make sure it looks right when making it bigger
* Make sure it looks right when making it smaller
* Start with a huge sprite and make it very small
* Start with a tiny sprite and make it tinier
* Start with a very small sprite and make it very big
* Edit a sprite while it is in different sizes
* Make the bounds of the sprite larger or smaller (erase parts)
* add graphics effects at different sizes
* rotate the sprite at different sizes

It's important to get Browser coverage for this issue
